### PR TITLE
[RFC] cmake: Add Cmake function for adding named sections to the linker file

### PR DESCRIPTION
**Pull request by oyvindronningstad**
Call add_named_section(".foo") to add a FLASH section with the name .foo.
Will also create the symbols _foo_start, _foo_end, and _foo_size.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

### CODE REVIEW

**Comment by oyvindronningstad** [comment]
cc @hakonfam @sigvartmh @osaether

File: cmake/extensions.cmake [hunk 1]
```diff
@@ -689,6 +689,80 @@ function(zephyr_check_compiler_flag lang option check)
...
+# To add c variables to the section:
+#   const mytype_t var_placed_in_foo _GENERIC_SECTION(.foo) __attribute__((used)) = something;
+#
+# Example 2:
+# add_named_section("bar", ROMABLE_REGION)
+# creates a section named bar in flash, and the symbols _bar_start, _bar_end,
+# and _bar_size.
+#
+set(generated_sections_filename "${PROJECT_BINARY_DIR}/include/generated/generated-sections.ld")
+file(WRITE ${generated_sections_filename} "")
```
Comment (SebastianBoe) [inline cmake/extensions.cmake]: So far, unless I've missed a patch, we have been able to limit this file to functions and macros. I'd like this to continue to hold if possible.

File: include/arch/arm/cortex_m/scripts/linker.ld [hunk 2]
```diff
@@ -515,6 +515,10 @@ SECTIONS
 #include <custom-sections.ld>
 #endif
 
+/* Located in generated directory. This file is populated by the
+ * add_named_section() Cmake function. */
+#include <generated-sections.ld>
+
```
Comment (SebastianBoe) [inline include/arch/arm/cortex_m/scripts/linker.ld]: We need to add this to all linker.ld unfortunately.

File: cmake/extensions.cmake (hunk 1 above)
Comment (oyvindronningstad) [reply cmake/extensions.cmake]: Done

File: include/arch/arm/cortex_m/scripts/linker.ld (hunk 2 above)
Comment (oyvindronningstad) [reply include/arch/arm/cortex_m/scripts/linker.ld]: Will do

**Comment by oyvindronningstad** [comment]
@SebastianBoe Thanks, I started addressing your comments. I ported CCM_*, TI_CCFG, and ramfunc so far. Take a look to see whether you think it's going in the right direction.

**Comment by SebastianBoe** [comment]
Can you elaborate on the motivation?

Is it because we think it is easier for users to interface with the add_named_section API than the LD API, or is it to support out-of-tree use-cases?

**Comment by oyvindronningstad** [comment]
> Can you elaborate on the motivation?

Yep, these are my main points:
 - Keep the linker definitions of a module together with everything else to do with that module.
 - Make it easier to add a section, syntax-wise.
 - Make it easier to enforce "best-practices", f.ex. the restriction on usage of `SECTION_DATA_PROLOGUE`
 - Make it less invasive to add a section, since you're less likely to interfere with what is already there. Right now it's hard/impossible to know if you are messing with the surrounding sections. It needs to be easier to see what sections/symbols must be placed in a specific position globally.
 - Make the linker script cleaner and more readable.

The out-of-tree point is a bonus, hadn't thought of that :).

**Comment by sigvartmh** [comment]
I really like this change as it cleans up the vendor specific stuff out of the linkerscript where it doesn't really belong.

**Comment by SebastianBoe** [comment]
I'll get back to this after multi-image is merged.

There are definitely some good ideas here that we want to take advantage of :)

File: soc/arm/st_stm32/common/CMakeLists.txt [hunk 3]
```diff
@@ -1,2 +1,8 @@
 zephyr_sources_ifdef(CONFIG_ARM_MPU arm_mpu_regions.c)
 zephyr_sources(stm32cube_hal.c)
+
+if (DEFINED DT_CCM_BASE_ADDRESS)
+  add_named_section(".ccm_bss" REGION CCM GROUP_START CCM)
```
Comment (SebastianBoe) [inline soc/arm/st_stm32/common/CMakeLists.txt]: Are you missing NOLOAD?

File: include/arch/arm/cortex_m/scripts/linker.ld [hunk 4]
```diff
@@ -472,45 +472,6 @@ SECTIONS
...
-	{
-		__ccm_data_start = .;
-		*(.ccm_data)
-		*(".ccm_data.*")
-	} GROUP_LINK_IN(CCM AT> ROMABLE_REGION)
-
-	__ccm_data_end = .;
-	__ccm_end = .;
-
-	__ccm_data_rom_start = LOADADDR(_CCM_DATA_SECTION_NAME);
```
Comment (SebastianBoe) [inline include/arch/arm/cortex_m/scripts/linker.ld]: Are we missing this symbol?

**Comment by SebastianBoe** [comment]
When I read this I am finding it easier to predict the behaviour of the original LD scripts than the new API.

My suggestion would therefore be to keep these improvements:

> Keep the linker definitions of a module together with everything else to do with that module.
> 
> Make it less invasive to add a section, since you're less likely to interfere with what is already there.
> 
> Right now it's hard/impossible to know if you are messing with the surrounding sections. It needs to be easier to see what sections/symbols must be placed in a specific position globally.
> 
> Make the linker script cleaner and more readable.
> 
> Support injecting multiple out-of-tree LD script snippets.

And leave these be as-is:

> Make it easier to add a section, syntax-wise.
> 
> Make it easier to enforce "best-practices", f.ex. the restriction on usage of SECTION_DATA_PROLOGUE
> 

by having the API append/concatenate local LD script snippets instead of generating LD script itself.

**Comment by oyvindronningstad** [comment]
> When I read this I am finding it easier to predict the behaviour of the original LD scripts than the new API.
> 
> My suggestion would therefore be to keep these improvements:
> 
> > Keep the linker definitions of a module together with everything else to do with that module.
> > Make it less invasive to add a section, since you're less likely to interfere with what is already there.
> > Right now it's hard/impossible to know if you are messing with the surrounding sections. It needs to be easier to see what sections/symbols must be placed in a specific position globally.
> > Make the linker script cleaner and more readable.
> > Support injecting multiple out-of-tree LD script snippets.
> 
> And leave these be as-is:
> 
> > Make it easier to add a section, syntax-wise.
> > Make it easier to enforce "best-practices", f.ex. the restriction on usage of SECTION_DATA_PROLOGUE
> 
> by having the API append/concatenate local LD script snippets instead of generating LD script itself.

I think this is a good call. I opened a separate pull request for the updated work (#109)

**Pull request closed without merging**